sway/desktop/output.c

```diff
@@ -243,28 +266,36 @@ static int output_repaint_timer_handler(void *data) {
 
 	output_configure_scene(output, &root->root_scene->tree.node, 1.0f);
 
-	struct wlr_scene_output_state_options opts = {
-		.color_transform = output->color_transform,
-	};
-
-	if (output->gamma_lut_changed) {
+	if (output->gamma_lut_changed || output->can_tear) {
 		struct wlr_output_state pending;
+		struct wlr_gamma_control_v1 *gamma_control = NULL;
+		bool tearing_state_updated = false;
+
 		wlr_output_state_init(&pending);
-		if (!wlr_scene_output_build_state(output->scene_output, &pending, &opts)) {
+		if (!wlr_scene_output_build_state(output->scene_output, &pending, NULL)) {
 			return 0;
 		}
 
-		output->gamma_lut_changed = false;
-		struct wlr_gamma_control_v1 *gamma_control =
-			wlr_gamma_control_manager_v1_get_control(
-			server.gamma_control_manager_v1, output->wlr_output);
-		if (!wlr_gamma_control_v1_apply(gamma_control, &pending)) {
-			wlr_output_state_finish(&pending);
-			return 0;
+		if (output->gamma_lut_changed) {
+			output->gamma_lut_changed = false;
+			gamma_control = wlr_gamma_control_manager_v1_get_control(
+				server.gamma_control_manager_v1, output->wlr_output);
+			if (!wlr_gamma_control_v1_apply(gamma_control, &pending)) {
+				wlr_output_state_finish(&pending);
+				return 0;
+			}
+		}
+
+		if (output->can_tear) {
+			pending.tearing_page_flip = true;
+			tearing_state_updated = true;
```

Comment:
Hm, it seems like these changes try to add tearing logic on top of the gamma LUT stuff? I don't think that works: if tearing fails the gamma LUT fails to apply, and the other way around. Moreover, if tearing fails there is no retry step.

Reply:
This might help a bit: https://github.com/swaywm/sway/commit/1e0031781fc9283db7096aba34deca5503c2ab91

Reply:
Just rebased my fork (https://github.com/vimproved/sway/tree/tearing_control) against https://github.com/swaywm/sway/commit/1e0031781fc9283db7096aba34deca5503c2ab91, so hopefully that should help.

> Moreover, if tearing fails there is no retry step.
Ah, seems I managed to accidentally get rid of the fallback somewhere in my fork. Will fix.

Reply:
Changed.

When the output state test fails, we now try again but with tearing page-flip disabled. I hope this is what you meant.

It works fine with atomic disabled, but with atomic enabled, the page-flip seems to always fail with tearing enabled:

```
00:00:14.731 [ERROR] [sway/desktop/output.c:299] Output test failed on 'DP-1', retrying without tearing page-flip
00:00:14.737 [DEBUG] [wlr] [backend/drm/atomic.c:79] connector DP-1: Atomic commit failed: Invalid argument
00:00:14.737 [DEBUG] [wlr] [backend/drm/atomic.c:84] (Atomic commit flags: PAGE_FLIP_ASYNC | ATOMIC_TEST_ONLY)
00:00:14.737 [ERROR] [sway/desktop/output.c:299] Output test failed on 'DP-1', retrying without tearing page-flip
00:00:14.743 [DEBUG] [wlr] [backend/drm/atomic.c:79] connector DP-1: Atomic commit failed: Invalid argument
00:00:14.743 [DEBUG] [wlr] [backend/drm/atomic.c:84] (Atomic commit flags: PAGE_FLIP_ASYNC | ATOMIC_TEST_ONLY)
```

Any idea why this happens, or what the status is in case it's a known issue?

Reply:
There are multiple issues with atomic on the kernel side. See:

- https://lore.kernel.org/dri-devel/20240629152204.666748-1-contact@emersion.fr/
- https://lore.kernel.org/dri-devel/20240702212215.109696-2-andrealmeid@igalia.com/

Reply:
The second patch was merged last week, but the first one still hasn't been despite getting a R-b over a month ago. Perhaps it fell through the cracks and needs a ping?

Reply:
v2 for second patch is here and waiting for feedback: https://lore.kernel.org/dri-devel/20240731191014.878320-1-contact@emersion.fr/